File: vpid_py/src/device_info.py
class USBDevice(object):
    def __init__(self, vid="", pid="", serial="", name="", 
                 manufacturer="", location="", driver="", 
                 device_id="", pnp_device_id="", status=""):
        self.vid = vid
        self.pid = pid
        self.serial = serial
        self.name = name
        self.manufacturer = manufacturer
        self.location = location
        self.driver = driver
        self.device_id = device_id
        self.pnp_device_id = pnp_device_id
        self.status = status

    def get_display_name(self):
        if self.name:
            return self.name
        if self.manufacturer:
            return "{0} Device".format(self.manufacturer)
        return "Unknown USB Device"

    def get_vid_pid_string(self):
        if self.vid and self.pid:
            return "VID: {0}  PID: {1}".format(self.vid, self.pid)
        return "VID: N/A  PID: N/A"

    def to_dict(self):
        return {
            "名称": self.name or "未知设备",
            "供应商ID (VID)": self.vid or "N/A",
            "产品ID (PID)": self.pid or "N/A",
            "序列号": self.serial or "N/A",
            "制造商": self.manufacturer or "N/A",
            "位置": self.location or "N/A",
            "驱动程序": self.driver or "N/A",
            "设备ID": self.device_id or "N/A",
            "状态": self.status or "N/A",
        }

    def to_clipboard_text(self):
        lines = [
            "名称: {0}".format(self.name or 'N/A'),
            "供应商ID (VID): {0}".format(self.vid or 'N/A'),
            "产品ID (PID): {0}".format(self.pid or 'N/A'),
            "序列号: {0}".format(self.serial or 'N/A'),
            "制造商: {0}".format(self.manufacturer or 'N/A'),
            "位置: {0}".format(self.location or 'N/A'),
            "驱动程序: {0}".format(self.driver or 'N/A'),
            "状态: {0}".format(self.status or 'N/A'),
        ]
        return "\n".join(lines)

File: native/src/core/device_info.cpp
#include "core/device_info.h"
#include <algorithm>
#include <cctype>

namespace vpid {

namespace {
// 去除 0x/0X 前缀并转大写
std::string strip0x(std::string s) {
    if (s.size() >= 2 && s[0] == '0' && (s[1] == 'x' || s[1] == 'X')) {
        s = s.substr(2);
    }
    std::transform(s.begin(), s.end(), s.begin(),
                   [](unsigned char c) { return (char)std::toupper((int)c); });
    return s;
}
} // namespace

std::string USBDevice::getDisplayName() const {
    if (!name.empty()) return name;
    if (!manufacturer.empty()) return manufacturer + " Device";
    return "Unknown USB Device";
}

std::string USBDevice::getFormattedVid() const {
    return vid.empty() ? std::string("N/A") : strip0x(vid);
}
std::string USBDevice::getFormattedPid() const {
    return pid.empty() ? std::string("N/A") : strip0x(pid);
}
std::string USBDevice::getVidPidString() const {
    return getFormattedVid() + ":" + getFormattedPid();
}
DeviceKey USBDevice::getUniqueKey() const { return {vid, pid, serial}; }

std::string USBDevice::toClipboardText() const {
    auto orN = [](const std::string& v)->std::string{ return v.empty() ? std::string("N/A") : v; };
    std::string n = name.empty() ? std::string("未知设备") : name;
    return "名称: "   + n + "\n"
         + "VID: "    + getFormattedVid() + "\n"
         + "PID: "    + getFormattedPid() + "\n"
         + "序列号: "  + orN(serial) + "\n"
         + "制造商: "  + orN(manufacturer) + "\n"
         + "位置: "    + orN(location) + "\n"
         + "路径: "    + orN(path) + "\n"
         + "驱动: "    + orN(driver) + "\n"
         + "状态: "    + orN(status);
}

} // namespace vpid